doc/comptime.org

```diff
@@ -0,0 +1,495 @@
+The results show that the comptime expression is faster than the runtime one.
+But this will work only with code that hasn't runtime dependency code.
+
+#+begin_src zig :imports '(std) :main 'no :testsuite 'yes :exports both
+  const Timer = std.time.Timer;
+  const stdout = std.io.getStdOut().writer();
+
+  fn fib(iteration: u32) u32 {
+      if (iteration < 2) return iteration;
+
+      return fib(iteration - 1) + fib(iteration - 2);
+  }
+
+  test "comptime fib" {
+      var timer = try Timer.start();
+      const result = comptime fib(15);
+      const elapsed = timer.read();
+      try stdout.print("Elasped-comptime: {d:0.2}ns\n", .{elapsed});
+
+      try std.testing.expectEqual(610, result);
+  }
+
+  test "fib" {
+      var timer = try Timer.start();
+      var result = fib(15);
+      const elapsed = timer.read();
+      try stdout.print("Elasped-runtime: {d:0.2}ns\n", .{elapsed});
+
+      try std.testing.expect(610 == result);
+  }
+
+#+end_src
+
+#+RESULTS:
+| Elasped-comptime: | 58ns   |
+| Elasped-runtime:  | 6078ns |
+
+*** Compile-time parameter
+Zig implements generic by using duck typing at compile-time.
+To use generic, Zig needs to know the type at compile-time.
+#+begin_src zig :imports '(std) :main 'no :testsuite 'yes
+
+  fn greater(comptime T: type, array: []const T) !?T {
+      var max: ?T = null;
+      for (array) |item| {
+          if (max) |m| {
+              if (m < item) {
+                  max = item;
+              }
+          } else {
+              max = item;
+          }
+      }
+      return max;
+  }
+
+  test "should return the max of an i32 array" {
+      const intArray = [_]i32{ 2, 9, 4, 6, 7, 1};
+      const result = try greater(i32, &intArray);
+
+      try std.testing.expect(result == 9);
+  }
+
+  test "should return the max of an f32 array" {
+      const floatArray = [_]f32{ 2.34, 14.55, 4.12, 6.876, 7.111 };
+      const result = try greater(f32, &floatArray);
+
+      try std.testing.expect(result == 14.55);
+  }
+
+
+#+end_src
+
+#+RESULTS:
+
+But with duck typing, if the same method is used, an error will be thrown at compile time:
+#+begin_src Zig
+  test "should fail with bool" {
+      const boolArray = [_]bool{ true, false, true, true };
+      const result = greater(bool, &boolArray);
+  }
+#+end_src
+
+The error will be:
+#+begin_example
+error: operator < not allowed for type 'bool'
+#+end_example
+
+Moreover, comptime can also be used as a type definition.
+For this, the function needs to return a ~type~.
+The example below is based on [[https://zig.guide/language-basics/comptime][the zig guide]] [cite:@ComptimeZigGuide2024], it's shows that it can define a new type with a function.
+#+begin_src zig :imports '(std) :main 'no :testsuite yes
+  fn Matrix(
+      comptime T: type,
+      comptime width: comptime_int,
+      comptime height: comptime_int,
+  ) type {
+      return [height][width]T;
+  }
+
+  fn Matrix3D(
```

Comment:
might lead to misunderstanding by calling this matrix "3D" since it is in fact 2D but has a height and lenght of 3